### PR TITLE
feat: add orderbook API base url

### PR DESCRIPTION
**Pull request by anxolin**
# Description
Allow to set your own Orderbook API URL instead of the default ones from the SDK.

This PR is a follow up on https://github.com/cowprotocol/cow-sdk/pull/177.

The motivation, is to allow to set our service IP instead of the external one, to make the connection more performant.

## Bonus
While testing this, i realized that in case of an error, the SDK re-attempts 10 times using exponential backoff. This makes it take too much time in case of an error. I reduced to 5 the attempts, and now it re-attempts, but it will take at most a few seconds instead of 2-3 minutes.

## Follow up
After this is merged, we will need to set it up (please, let me do it!) 

See
https://cowservices.slack.com/archives/C05RMJB7ZFA/p1697021044757619?thread_ts=1697007472.259279&cid=C05RMJB7ZFA

## Test
Pass the `--orderBookApi ` and make sure it uses the parameter

```bash
ts-node ./src/index.ts run --rpc <your-rpc> --dployment-block 9847564 --page-size 5000 --orderBookApi https://api.cow.fi/goerli
```

Once this is in staging we can test this in the infrastructure project.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -100,6 +100,16 @@ const deploymentBlockOption = new Option(
...
 
+const multiOrderBookApiOption = new Option(
+  "--orderBookApi <orderBookApi...>",
+  "Orderbook API base URLs (i.e. https://api.cow.fi/mainnet, https://api.cow.fi/xdai, etc.)"
+).default([]);
+
+const orderBookApiOption = new Option(
+  "--orderBookApi <orderBookApi>",
+  "Orderbook API base URL (i.e. https://api.cow.fi/mainnet)"
+).default(undefined);
```
Comment (mfw78) [inline src/index.ts:111]: As this option doesn't have "make mandatory", is setting the default to `undefined` required?

File: src/index.ts [hunk 2]
```diff
@@ -119,7 +130,8 @@ async function main() {
     .addOption(disableNotificationsOption)
     .addOption(slackWebhookOption)
     .action((options) => {
-      const { logLevel } = options;
+      const { logLevel, orderBookApi } = options;
+
```
Comment (mfw78) [inline src/index.ts:134]: As the `orderBookApi` isn't being modified here, or checked for any sanity, is it required to be deconstructed here? It should just pass in with the `...options` below, no?

File: src/domain/chainContext.ts [hunk 3]
```diff
@@ -120,7 +140,13 @@ export class ChainContext {
...
 
     // Save the context to the static map to be used by the API
-    const context = new ChainContext(options, provider, chainId, registry);
+    const context = new ChainContext(
+      options,
+      provider,
+      chainId,
+      registry,
+      orderBookApi
+    );
```
Comment (mfw78) [inline src/domain/chainContext.ts:149]: As `orderBookApi` is part of the `options`, would we not just use the `options` within the constructor as this is basically already passed the whole way down and we can DRY this?

**Review by mfw78 — APPROVED** [review]
Great that we can configure this now. I think there's some scope for some minor DRY'ing.

File: src/index.ts (hunk 2 above)
Comment (anxolin) [reply src/index.ts:134]: 👍

File: src/index.ts (hunk 1 above)
Comment (anxolin) [reply src/index.ts:111]: https://github.com/cowprotocol/watch-tower/pull/106#event-10687336475

File: src/index.ts (hunk 2 above)
Comment (anxolin) [reply src/index.ts:134]: https://github.com/cowprotocol/watch-tower/pull/115

**Pull request merged**